src/darwin/Framework/CHIP/ServerEndpoint/MTRServerAccessControl.mm

```diff
@@ -0,0 +1,193 @@
+            case MTRAccessControlEntryPrivilegeManage:
+                if (requestPrivilege == Privilege::kView || requestPrivilege == Privilege::kOperate || requestPrivilege == Privilege::kManage) {
+                    return CHIP_NO_ERROR;
+                }
+                break;
+            case MTRAccessControlEntryPrivilegeAdminister:
+                if (requestPrivilege == Privilege::kView || requestPrivilege == Privilege::kProxyView || requestPrivilege == Privilege::kOperate || requestPrivilege == Privilege::kManage || requestPrivilege == Privilege::kAdminister) {
+                    return CHIP_NO_ERROR;
+                }
+                break;
+            default:
+                MTR_LOG_ERROR("Uknown granted privilege %u, ignoring", grant.grantedPrivilege);
+                break;
+            }
+
+            // If this grant did not match, just move on to the next one.
+        }
+
+        // None of the grants matched.
+        return CHIP_ERROR_ACCESS_DENIED;
+    }
+
+    bool GrantSubjectMatchesDescriptor(MTRAccessGrant * grant, const SubjectDescriptor & descriptor)
+    {
+        if (grant.subjectID == nil) {
+            // This is an all-nodes grant for CASE access only.
+            return descriptor.authMode == AuthMode::kCase;
+        }
+
+        NodeId grantSubjectNodeId = grant.subjectID.unsignedLongLongValue;
+        if (IsOperationalNodeId(grantSubjectNodeId)) {
+            return descriptor.authMode == AuthMode::kCase && descriptor.subject == grantSubjectNodeId;
+        }
+
+        if (IsGroupId(grantSubjectNodeId)) {
+            return descriptor.authMode == AuthMode::kGroup && descriptor.subject == grantSubjectNodeId;
+        }
+
+        if (IsCASEAuthTag(grantSubjectNodeId)) {
+            return descriptor.cats.CheckSubjectAgainstCATs(grantSubjectNodeId);
+        }
+
+        MTR_LOG_ERROR("Unexpected grant subject: 0x%llx", grantSubjectNodeId);
+        return false;
+    }
+};
+
+struct ControllerAccessControl {
+    DeviceTypeResolver mDeviceTypeResolver;
+    AccessControlDelegate mDelegate;
+    ControllerAccessControl() { GetAccessControl().Init(&mDelegate, mDeviceTypeResolver); }
+};
+
+Global<ControllerAccessControl> gControllerAccessControl;
+
+} // anonymous namespace
+
+int MatterGetAccessPrivilegeForReadEvent(ClusterId cluster, EventId event)
+{
+    // We don't support any event bits yet.
+    return kMatterAccessPrivilegeAdminister;
+}
+
+int MatterGetAccessPrivilegeForInvokeCommand(ClusterId cluster, CommandId command)
+{
+    // For now we only have OTA, which uses Operate.
+    return kMatterAccessPrivilegeOperate;
+}
+
+int MatterGetAccessPrivilegeForReadAttribute(ClusterId cluster, AttributeId attribute)
+{
+    NSNumber * _Nullable neededPrivilege = [[MTRDeviceControllerFactory sharedInstance] neededReadPrivilegeForClusterID:@(cluster) attributeID:@(attribute)];
+    if (neededPrivilege == nil) {
+        // Default is View.
+        return kMatterAccessPrivilegeView;
```

Comment:
Where does this default come from? Maybe by your logic for unknown values this should also be Admin?

Reply:
Default comes from spec, sort of.  But you are right; probably better to fail closed here.  Will fix.

Reply:
Done.